nixos/modules/services/network-filesystems/iroh.nix

```diff
@@ -0,0 +1,275 @@
+{ config, lib, pkgs, utils, ... }:
+with lib;
+let
+  # NB: build the manual with:
+  # $ nix-build nixos/release.nix -A manual.x86_64-linux
+
+  # NB: run the tests with:
+  # $ nix-build -A nixosTests.iroh
+
+  # Docs for writing tests:
+  # https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests
+
+  # TODO test: configure systemd logs directory:
+  # iroh: Failed to create log directory: Read-only filesystem
+
+  # TODO run systemd-analyze security and apply whatever sandboxing doesn't break the service
+  # TODO add tests
+
+  # cfg is a conventional alias for the configuration of the current module
+  cfg = config.services.iroh;
+
+  # Default listening port.
+  listenPort = 11204;
+  # Default RPC port.
+  rpcPort = 1337;
+  # Default metrics port. With a value of -1 serving metrics is disabled.
+  metricsPort = -1;
+
+  # --add <ADD>: optionally add a file or folder to the node
+  irohFlags = utils.escapeSystemdExecArgs (
+    optional (cfg.metricsPort != -1) "--metrics-port ${cfg.metricsPort}" ++
+    optional cfg.noTicket "--no-ticket" ++
+    optional (cfg.tag != null) "--tag ${cfg.tag}" ++
+    cfg.extraFlags
+  );
+
+  # A submodule to describe where logs may be placed.
+  logOutput = types.submodule {
+    options = {
+      # TODO output should be: rust=$level (a tracing_subscriber::EnvFilter)
+      level = mkOption {
+        type = types.enum [ "error" "warn" "info" "debug" "trace" ];
+        default = "info";
+      };
+      rotation = mkOption {
+        type = types.enum [ "hourly" "daily" "never" ];
+        default = "daily";
+      };
+      maxFiles = mkOption {
+        # TODO make sure unsigned >= 1
+        type = types.int;
+        default = 1;
+      };
+      dir = mkOption {
+        type = types.path;
+      };
+    };
+  };
+
+  # A submodule that describes a relay node used to assist in NAT traversal.
+  relayNode = types.submodule {
+    options = {
+      # TODO use regex to match URL, including https:// prefix!
+      url = mkOption {
+        description = "The URL of the relay node.";
+        type = types.str;
+      };
+
+      stunOnly = mkOption {
+        description = "Only use STUN for traversal of NAT gateways.";
+        type = types.bool;
+      };
+
+      stunPort = mkOption {
+        description = "Port at which STUN service is listening.";
+        type = types.port;
+      };
+    };
+  };
+
+  # The total set of relays is the merge of relayNodes (which are the default nodes provided by
+  # iroh, if not replaced) and any extra nodes defined by the user.
+  relaySettings = cfg.relayNodes // cfg.extraRelayNodes;
+
+  # The config file should be located in the data directory and is NOT created by default. All
+  # values are optional, including the file itself.
+  configFormat = pkgs.formats.toml { };
+  #configFile = configFormat.generate "${cfg.dataDir}/iroh.config.toml" relaySettings;
+  configFile = configFormat.generate "/var/lib/iroh/iroh.config.toml" relaySettings;
```

Comment:
this won't place a file under `/var/lib/iroh`

does `iroh` accept an immutable config file and directory? if so we should use the `IROH_CONFIG_DIR` environment variable to create a directory of configuration files...

maybe something like this: 

`environment.etc."iroh/iroh.config.toml".source = configFormat.generate "iroh.config.toml" relaySettings;`